tests/test_butler.py

```diff
@@ -0,0 +1,93 @@
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Output butler from pipelines_check test run."""
+
+import os
+import unittest
+
+from lsst.daf.butler import Butler
+
+TESTDIR = os.path.abspath(os.path.dirname(__file__))
+
+MAIN_CHAIN = "demo_collection"
+EXE_CHAIN = "demo_collection_exe"
+
+
+class PiplinesCheckTestCase(unittest.TestCase):
+    """Check outputs from test run."""
+
+    def setUp(self):
+        """Create a new butler root for each test."""
+        root = os.path.join(TESTDIR, os.path.pardir, "DATA_REPO")
+        self.butler = Butler(root, writeable=False)
+
+    def _get_datasets_from_chain(self, chain, datasetType=...):
+        """Return all the datasets from the first run in chain.
+
+        Datasets are all unresolved.
+        """
+        collections = list(self.butler.registry.queryCollections(chain, flattenChains=True))
+        run = collections.pop(0)
+        print(f"Retrieving datasets from run {run}")
+
+        unresolved = {ref.unresolved() for ref in self.butler.registry.queryDatasets(datasetType=datasetType,
+                                                                                     collections=run)}
+        return unresolved
+
+    def testExecutionButler(self):
+        """Check outputs match in both runs."""
+
+        # Check that we have identical datasets in both collections
+        # except for the dataset.id
+        main_datasets = self._get_datasets_from_chain(MAIN_CHAIN)
+        exe_datasets = self._get_datasets_from_chain(EXE_CHAIN)
+        self.assertGreater(len(exe_datasets), 0)
+        self.assertEqual(len(main_datasets), len(exe_datasets))
+
+        difference = main_datasets - exe_datasets
+        self.assertEqual(len(difference), 0, "Some datasets missing.")
+
+    def testExecutionExistence(self):
+        """Check that the execution butler files are really there."""
+
+        exe_datasets = self._get_datasets_from_chain(EXE_CHAIN)
+        for ref in exe_datasets:
+            self.assertTrue(self.butler.datasetExists(ref, collections=EXE_CHAIN))
+
+    def testLogDataset(self):
+        """Ensure that the logs are captured in both modes."""
+
+        log_datasets = self._get_datasets_from_chain(MAIN_CHAIN, datasetType="isr_log")
+        self.assertEqual(len(log_datasets), 1)
+
+        isr_log_ref = log_datasets.pop()
+
+        # Get the logs from both main and exe collections.
+        main_isr_log = self.butler.get("isr_log", dataId=isr_log_ref.dataId, collections=MAIN_CHAIN)
+        exe_isr_log = self.butler.get("isr_log", dataId=isr_log_ref.dataId, collections=EXE_CHAIN)
+
+        print(exe_isr_log)
```

Comment:
Is this a debugging artifact?

Reply:
Yes and no. I thought it would be nice to show it if there was a test failure. It's not doing any harm there since it's captured by pytest and only displayed on test fail.

Reply:
I put the log printing into the failure message itself.